File: sdks/nodejs/src/emails.ts
import { CloudPoolClient } from './index';

export interface SendEmailParams {
  to: string | string[];
  subject: string;
  body: string;
  from?: string;
  cc?: string | string[];
  bcc?: string | string[];
  contentType?: 'text/plain' | 'text/html';
}

export class EmailsClient {
  constructor(private readonly client: CloudPoolClient) {}

  public async send<T = unknown>(params: SendEmailParams): Promise<T> {
    const response = await this.client.request(
      'POST',
      '/api/v1/emails/send',
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(params),
      }
    );
    return response.json();
  }

  public async listInbox<T = unknown>(params?: {
    limit?: number;
    unreadOnly?: boolean;
  }): Promise<T[]> {
    const searchParams = new URLSearchParams();
    if (params?.limit) searchParams.set('limit', String(params.limit));
    if (params?.unreadOnly)
      searchParams.set('unreadOnly', 'true');
    const query = searchParams.toString();
    const response = await this.client.request(
      'GET',
      `/api/v1/emails/inbox${query ? '?' + query : ''}`
    );
    return response.json();
  }
}

File: sdks/rust/src/emails.rs
use reqwest::Method;
use serde_json::Value;

use crate::client::CloudPoolClient;
use crate::errors::CloudPoolError;

/// Client for sending emails and reading the inbox.
pub struct EmailsClient<'a> {
    client: &'a CloudPoolClient,
}

impl<'a> EmailsClient<'a> {
    /// Create a new emails client.
    pub fn new(client: &'a CloudPoolClient) -> Self {
        Self { client }
    }

    /// Send an email.
    pub async fn send(&self, to: &str, subject: &str, body: &str) -> Result<Value, CloudPoolError> {
        let payload = serde_json::json!({
            "to": to,
            "subject": subject,
            "body": body,
        });
        let res = self
            .client
            .send_with_retry(self.client.request(Method::POST, "emails/send").json(&payload))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List inbox messages.
    pub async fn list_inbox(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "emails/inbox"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }
}
